### PR TITLE
handle legacy GHE servers

### PR DESCRIPTION
**Pull request by bcanzanella**
This PR fixes several issues around older versions of GitHub Enterprise.

https://trello.com/c/h4k4YVhv/5110-endless-loading-trying-to-view-a-pr
(related to unsupported `IsDraft` property)

https://trello.com/c/6gu4kw6g/4976-error-in-gh-pr-query-related-do-nonexistent-isdraft-field
(unsupported `IsDraft` property)

https://trello.com/c/txBeikS2/5129-pass-ghe-pr-id-back-after-creation
have a better UX, by allowing the user to immediately go to the CS PR panel. before, we weren't capturing the PR's id. <div>
https://trello.com/c/aMaUjLlc/5130-pr-list-isnt-updated-after-a-pr-change-add-close
if you add a PR or close a PR, the sidebar's open pull requests view was not updating.
<div><br></div><div>https://trello.com/c/Xl8DQzGo/5140-parse-error-on-colon</div><div><br></div><div>The way i got around the non-existent properties is using a tokenization system that takes into account the current api version, it uses a syntax like `[theProp:&gt;X.Y.Z]` which will get transformed into `theProp` or removed. thus, these props are nullable.</div><div><br></div><div>Also, the older GHE instances require the use of `ID` types instead of `String`s (newer GHE instances are more forgiving), so there are a lot of changes w/r/t this.</div></div>

### CODE REVIEW

File: shared/agent/src/providers/github.ts [hunk 1]
```diff
@@ -1644,15 +1719,15 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
...
+						login
+						avatarUrl(size: 20)
+						url
 					}
 					body
 					bodyText
 					number
 					state
-					isDraft
+					${this._transform(`[isDraft:>=2.21.0]`)}
```
Comment (marcelofarias) [inline shared/agent/src/providers/github.ts:1730]: love it

File: shared/agent/src/providers/github.ts [hunk 2]
```diff
@@ -1891,28 +1984,25 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
...
+		const query = `mutation MergePullRequest($pullRequestId:ID!) {
 			mergePullRequest(input: {pullRequestId: $pullRequestId, mergeMethod: ${request.mergeMethod}}) {
 				  clientMutationId
 				}
 			  }`;
 
-		const response = await this.mutate<any>(query, {
-			pullRequestId: request.pullRequestId
-		});
 		return true;
```
Comment (marcelofarias) [inline shared/agent/src/providers/github.ts:1993]: I believe we are missing the actual API call here

File: shared/agent/src/providers/github.ts [hunk 3]
```diff
@@ -2180,16 +2285,12 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
...
 				addComment(input: {subjectId: $subjectId, body:$body}) {
 				  clientMutationId
 				}
 			  }`;
 
-		const response = await this.mutate<any>(query, {
-			subjectId: request.pullRequestId,
-			body: request.text
-		});
 		return true;
```
Comment (marcelofarias) [inline shared/agent/src/providers/github.ts:2294]: We might be missing the actual API call here as well

**Review by marcelofarias — CHANGES_REQUESTED** [review]
I think you inadvertently removed a couple `transform` calls

**Comment by bcanzanella** [comment]
&gt; I think you inadvertently removed a couple `transform` calls<br>I sure did... fixing now.

**Comment by bcanzanella** [comment]
can you take a final look at this @mfarias&nbsp;?

**Review by marcelofarias — APPROVED** [review]

File: shared/agent/src/providers/github.ts [hunk 4]
```diff
@@ -688,7 +701,7 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
 			}
 
 			const createPullRequestResponse = await this.mutate<GitHubCreatePullRequestResponse>(
-				`mutation CreatePullRequest($repositoryId:String!, $baseRefName:String!, $headRefName:String!, $title:String!, $body:String!) {
+				`mutation CreatePullRequest($repositoryId:ID!, $baseRefName:String!, $headRefName:String!, $title:String!, $body:String!) {
 					__typename
 					createPullRequest(input: {repositoryId: $repositoryId, baseRefName: $baseRefName, headRefName: $headRefName, title: $title, body: $body}) {
 					  pullRequest {
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:707]: 1

File: shared/agent/src/providers/github.ts [hunk 5]
```diff
@@ -2199,13 +2309,15 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
...
 			pullRequestId: request.pullRequestId,
 			sha: request.rightSha,
 			text: request.text || "",
 			path: request.filePath,
-			startLine: request.startLine
+			startLine: request.startLine,
+			position: request.position
 		});
 
 		this._pullRequestCache.delete(request.pullRequestId);
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:2323]: 2

**Comment by bcanzanella** [comment]
1

```
const query = `query pr($owner:String!, $repo:String!) {
				repository(name: $repo, owner: $owner${cursor ? `, after: $cursor` : ""}) {
					pullRequests(states: [OPEN, MERGED], first: 100, orderBy: { field: UPDATED_AT, direction: DESC }) {
						totalCount
						pageInfo {
							startCursor
							endCursor
							hasNextPage
```
								
[shared/agent/src/providers/github.ts](https://github.com/teamcodestream/codestream/blob/595c2878bcee4e113dc778e802ec5b6e926b9ac3/shared/agent/src/providers/github.ts#L3500-L3507) (Lines 3500-3507)

File: shared/agent/src/providers/github.ts (hunk 4 above)
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:707]: a

File: shared/agent/src/providers/github.ts [hunk 6]
```diff
@@ -688,7 +701,7 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
 			}
 
 			const createPullRequestResponse = await this.mutate<GitHubCreatePullRequestResponse>(
-				`mutation CreatePullRequest($repositoryId:String!, $baseRefName:String!, $headRefName:String!, $title:String!, $body:String!) {
+				`mutation CreatePullRequest($repositoryId:ID!, $baseRefName:String!, $headRefName:String!, $title:String!, $body:String!) {
 					__typename
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:705]: 1

**Comment by bcanzanella** [comment]
2

```
g()
	getPullRequestDocumentMarkers({
		uri,
		repoId,
		streamId
	}: {
```
								
[shared/agent/src/providers/github.ts](https://github.com/teamcodestream/codestream/blob/595c2878bcee4e113dc778e802ec5b6e926b9ac3/shared/agent/src/providers/github.ts#L667-L672) (Lines 667-672)

File: shared/agent/src/providers/github.ts [hunk 7]
```diff
@@ -172,8 +179,8 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
 		restApi: { fns: {} }
 	};
 
-	async query<T = any>(query: string, variables: any = undefined) {
-		const response = await this.client.request<any>(query, variables);
+	async query(query: string, variables: any = undefined) {
+		const response = await (await this.client()).request<any>(query, variables);
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:184]: 3 empty lines around

File: shared/agent/src/providers/github.ts [hunk 8]
```diff
@@ -194,7 +201,10 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
 						functionName = e.stack
 							.split("\n")
 							.filter(
-								_ => _.indexOf("GitHubProvider") > -1 && _.indexOf("GitHubProvider.query") === -1
+								_ =>
+									(_.indexOf("GitHubProvider") > -1 ||
+										_.indexOf("GitHubEnterpriseProvider") > -1) &&
+									_.indexOf(".query") === -1
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:207]: 4

File: shared/agent/src/providers/github.ts [hunk 9]
```diff
@@ -135,8 +130,17 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
...
 	}
 
-	private _client: GraphQLClient | undefined;
-	protected get client(): GraphQLClient {
+	/**
+	 * The version of the GitHub api... note this is only set for enterprise
+	 *
+	 * @protected
+	 * @type {(string | undefined)}
+	 * @memberof GitHubProvider
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:138]: 5

File: shared/agent/src/providers/github.ts [hunk 10]
```diff
@@ -40,12 +39,8 @@ import {
...
 
-import {
-	CodemarkType,
-	CSGitHubProviderInfo,
-	CSReferenceLocation,
-	CSRepository
-} from "../protocol/api.protocol";
+import semver from "semver";
+import { CSGitHubProviderInfo, CSRepository } from "../protocol/api.protocol";
 import { Arrays, Functions, log, lspProvider, Strings } from "../system";
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:44]: 6

File: shared/agent/src/providers/github.ts (hunk 4 above)
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:707]: a

File: shared/agent/src/providers/github.ts (hunk 6 above)
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:705]: a

**Comment by bcanzanella** [comment]
a

```
	`mutation CreatePullRequest($repositoryId:ID!, $baseRefName:String!, $headRefName:String!, $title:String!, $body:String!) {
					__typename
					createPullRequest(input: {repositoryId: $repositoryId, baseRefName: $baseRefName, headRefName: $headRefName, title: $title, body: $body}) {
					  pullRequest {
							number,
							id,
							url,
							title
						}
```
								
[shared/agent/src/providers/github.ts](https://github.com/teamcodestream/codestream/blob/595c2878bcee4e113dc778e802ec5b6e926b9ac3/shared/agent/src/providers/github.ts#L704-L712) (Lines 704-712)

**Comment by bcanzanella** [comment]
4

```
`mutation CreatePullRequest($repositoryId:ID!, $baseRefName:String!, $headRefName:String!, $title:String!, $body:String!) {
					__typename
					createPullRequest(input: {repositoryId: $repositoryId, baseRefName: $baseRefName, headRefName: $headRefName, title: $title, body: $body}) {
					  pullRequest {
							number,
							id,
							url,
							title
						}
					}
```
								
[shared/agent/src/providers/github.ts](https://github.com/teamcodestream/codestream/blob/595c2878bcee4e113dc778e802ec5b6e926b9ac3/shared/agent/src/providers/github.ts#L704-L713) (Lines 704-713)

File: shared/agent/src/providers/github.ts (hunk 6 above)
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:705]: 2

File: shared/agent/src/providers/github.ts [hunk 11]
```diff
@@ -1219,7 +1232,7 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
 			});
 			return response;
 		} else {
-			// const query = `mutation UpdateDraft($pullRequestId: String!, $isDraft:Boolean!) {
+			// const query = `mutation UpdateDraft($pullRequestId:ID!, $isDraft:Boolean!) {
 			// 	updatePullRequest(input: {pullRequestId: $pullRequestId, isDraft:$isDraft}) {
 			// 		  clientMutationId
 			// 		}
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:1238]: a

File: shared/agent/src/providers/github.ts [hunk 12]
```diff
@@ -1219,7 +1232,7 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
 			});
 			return response;
 		} else {
-			// const query = `mutation UpdateDraft($pullRequestId: String!, $isDraft:Boolean!) {
+			// const query = `mutation UpdateDraft($pullRequestId:ID!, $isDraft:Boolean!) {
 			// 	updatePullRequest(input: {pullRequestId: $pullRequestId, isDraft:$isDraft}) {
 			// 		  clientMutationId
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:1237]: a

File: shared/agent/src/providers/github.ts [hunk 13]
```diff
@@ -1208,7 +1221,7 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
 
 	async setIsDraftPullRequest(request: { pullRequestId: string; isDraft: boolean }) {
 		if (!request.isDraft) {
-			const query = `mutation MarkPullRequestReadyForReview($pullRequestId: String!) {
+			const query = `mutation MarkPullRequestReadyForReview($pullRequestId:ID!) {
 				markPullRequestReadyForReview(input: {pullRequestId: $pullRequestId}) {
 					  clientMutationId
 					}
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:1227]: a

File: shared/agent/src/providers/github.ts [hunk 14]
```diff
@@ -1219,7 +1232,7 @@ export class GitHubProvider extends ThirdPartyIssueProviderBase<CSGitHubProvider
 			});
 			return response;
 		} else {
-			// const query = `mutation UpdateDraft($pullRequestId: String!, $isDraft:Boolean!) {
+			// const query = `mutation UpdateDraft($pullRequestId:ID!, $isDraft:Boolean!) {
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:1235]: 1

File: shared/agent/src/providers/github.ts (hunk 11 above)
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:1238]: g

Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:1238]: a

Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:1238]: a

File: shared/agent/src/providers/github.ts [hunk 15]
```diff
@@ -6,8 +6,7 @@ import * as paths from "path";
 import * as qs from "querystring";
 import { CodeStreamSession } from "session";
 import { URI } from "vscode-uri";
-import { Container, SessionContainer } from "../container";
-import { MarkerLocation } from "../api/extensions";
+import { SessionContainer } from "../container";
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:9]: a

File: shared/agent/src/providers/github.ts [hunk 16]
```diff
@@ -40,12 +39,8 @@ import {
...
 } from "../protocol/agent.protocol";
 
-import {
-	CodemarkType,
-	CSGitHubProviderInfo,
-	CSReferenceLocation,
-	CSRepository
-} from "../protocol/api.protocol";
+import semver from "semver";
+import { CSGitHubProviderInfo, CSRepository } from "../protocol/api.protocol";
```
Comment (bcanzanella) [inline shared/agent/src/providers/github.ts:43]: a

**Pull request merged**